bsp/Src/I2C.c

```diff
@@ -0,0 +1,442 @@
+		// Check if the queue was created successfully
+		if (I2C2_Queue == NULL)
+			// Handle queue creation failure
+			return I2C_QFAIL;
+
+		HAL_NVIC_SetPriority(I2C2_EV_IRQn , I2C2PRIO, I2C2SPRIO); // set to priority (default 5) (not the highest priority) for I2C peripheral's interrupt
+		HAL_NVIC_EnableIRQ(I2C2_EV_IRQn); // Enable the I2C interrupt
+		HAL_NVIC_SetPriority(I2C2_ER_IRQn , I2C2PRIO, I2C2SPRIO); // set to priority (default 5) (not the highest priority) for I2C peripheral's interrupt
+		HAL_NVIC_EnableIRQ(I2C2_ER_IRQn); // Enable the I2C interrupt
+		if (HAL_I2C_Init(hi2c) != HAL_OK)
+		{
+			// Hancle I2C initialization error
+			return I2C_IFAIL;
+		}
+	}
+	#endif
+
+	#ifdef I2C3
+	else if(hi2c->Instance == I2C3)
+	{
+		__HAL_RCC_I2C3_CLK_ENABLE();
+		I2C3_Queue = xQueueCreateStatic(I2C3_QUEUE_SIZE, sizeof(uint8_t), I2C3_DataStore, &I2C3_DataQueue);
+		// Check if the queue was created successfully
+		if (I2C3_Queue == NULL)
+			// Handle queue creation failure
+			return I2C_QFAIL;
+
+		HAL_NVIC_SetPriority(I2C3_EV_IRQn , I2C3PRIO, I2C3SPRIO); // set to priority (default 5) (not the highest priority) for I2C peripheral's interrupt
+		HAL_NVIC_EnableIRQ(I2C3_EV_IRQn); // Enable the I2C interrupt
+		HAL_NVIC_SetPriority(I2C3_ER_IRQn , I2C3PRIO, I2C3SPRIO); // set to priority (default 5) (not the highest priority) for I2C peripheral's interrupt
+		HAL_NVIC_EnableIRQ(I2C3_ER_IRQn); // Enable the I2C interrupt
+		if (HAL_I2C_Init(hi2c) != HAL_OK)
+		{
+			// Hancle I2C initialization error
+			return I2C_IFAIL;
+		}
+	}
+	#endif
+
+	return I2C_OK;
+}
+
+/**
+* @brief    Transmits data onto the I2C bus.
+* @param    hi2c           to structure that has config data
+* @param    deviceAddr     target device address
+* @param    pDataBuff      data buffer
+* @param    len            amount of data
+* @return   I2C_StatusTypeDef
+*/
+I2C_StatusTypeDef i2c_send(I2C_HandleTypeDef *hi2c,
+              uint8_t deviceAddr,
+              uint8_t* pDataBuff,
+              uint16_t len)
+{
+	// Confirm length is not larger than its designated payload size
+	if(hi2c->Instance == I2C1 && len > I2C1_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+	else if(hi2c->Instance == I2C2 && len > I2C2_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+	else if(hi2c->Instance == I2C3 && len > I2C3_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+
+	// Update the latest I2C peripheral
+	update_last_i2c(hi2c);
+
+	if(HAL_I2C_GetState(hi2c) == HAL_I2C_STATE_READY)
+	{
+		return I2C_ConvertHALStatus(HAL_I2C_Master_Transmit_IT(hi2c, deviceAddr, pDataBuff, len));
+	}
+	
+	QueueHandle_t I2C_Queue = I2C1_Queue;
+	if(hi2c->Instance == I2C1) {}
+	else if (hi2c->Instance == I2C2)
+	{
+		I2C_Queue = I2C2_Queue;
+	}
+	else if (hi2c->Instance == I2C3)
+	{
+		I2C_Queue = I2C3_Queue;
+	}
+	else
+	{
+		// Handle invalid I2C peripheral
+		return I2C_BADINST;
+	}
+
+	if (len > uxQueueSpacesAvailable(I2C_Queue))
+	{
+		return I2C_QEMPTY;

```

Comment:
queue full?

Reply:
Call me a boeing 747 the way my head is in the clouds rn

Reply:
I've removed the QEMPTY error code because it should never happen